ambry-store/src/main/java/com/github/ambry/store/PersistentIndex.java

```diff
@@ -980,35 +964,54 @@ IndexValue markAsPermanent(StoreKey id, FileSpan fileSpan, MessageInfo info, lon
    * @throws StoreException if there is any problem writing the index record
    */
   IndexValue markAsUndeleted(StoreKey id, FileSpan fileSpan, long operationTimeMs) throws StoreException {
-    return markAsUndeleted(id, fileSpan, operationTimeMs, MessageInfo.LIFE_VERSION_FROM_FRONTEND);
+    return markAsUndeleted(id, fileSpan, null, operationTimeMs, MessageInfo.LIFE_VERSION_FROM_FRONTEND);
   }
 
   /**
    * Marks a blob as undeleted
    * @param id the {@link StoreKey} of the blob
    * @param fileSpan the file span represented by this entry in the log
    * @param operationTimeMs the time of the update operation
+   * @param info this needs to be non-null in the case of recovery and replicateion. Can be {@code null} otherwise. Used if the PUT
+   *             record could not be found
    * @param lifeVersion lifeVersion of this undelete record.
    * @return the {@link IndexValue} of the undelete record
    * @throws StoreException if there is any problem writing the index record
    */
-  IndexValue markAsUndeleted(StoreKey id, FileSpan fileSpan, long operationTimeMs, short lifeVersion)
+  IndexValue markAsUndeleted(StoreKey id, FileSpan fileSpan, MessageInfo info, long operationTimeMs, short lifeVersion)
       throws StoreException {
     boolean hasLifeVersion = IndexValue.hasLifeVersion(lifeVersion);
     validateFileSpan(fileSpan, true);
+    long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
     List<IndexValue> values =
         findAllIndexValuesForKey(id, null, EnumSet.allOf(IndexEntryType.class), validIndexSegments);
-    validateSanityForUndelete(id, values, lifeVersion);
-    // This value is the delete IndexValue
-    IndexValue value = values.get(0);
-    maybeChangeExpirationDate(value, values);
-    lifeVersion = hasLifeVersion ? lifeVersion : (short) (value.getLifeVersion() + 1);
-    long size = fileSpan.getEndOffset().getOffset() - fileSpan.getStartOffset().getOffset();
-    IndexValue newValue =
-        new IndexValue(value.getSize(), value.getOffset(), value.getFlags(), value.getExpiresAtMs(), operationTimeMs,
-            value.getAccountId(), value.getContainerId(), lifeVersion);
-    newValue.setNewOffset(fileSpan.getStartOffset());
-    newValue.setNewSize(size);
+    IndexValue newValue;
+    if (info != null) {
+      // This is from recovery. In recovery, we don't need to do any sanity check because
+      // 1. we already know the IndexValue has it's source in the log.
+      // 2. some sanity check will fail.
+      //    assume we have P, D, U, D in the log, then a compaction cycle compacted P and first D,
```

Comment:
In this case, we still recovery Undelete index, but P can't be accessed, right?

Reply:
That's right. We used to think that when there is a U, there must be a P before U. This is wrong. After compaction, anything could happen. After compaction, in this case, we are left with only U and D, without any P. And recovery need to work for this case as well.